### ISSUE TITLE
[HPOS] Replace `get_post_type()` with WC_Data_Store::load( 'subscription' )->get_order_type( $id )

### ISSUE THREAD

**Issue by mattallan** [Automattic/woocommerce-subscriptions-core#288]
## Describe the bug

To support subscriptions stored in different data stores, we need to replace all cases of `get_post_type()` with `WC_Data_Store::load( 'subscription' )->get_order_type( $id )`.

Doing a search of `get_post_type(` in `subscriptions-core` code there's 15 instances of this still being used.

In #153 we only targetted code which used this function on WC Orders, but as part of tackling this issue we should replace all cases.

---

#### Updating `get_post_type( $subscription_id )`

```
'shop_subscription' == get_post_type( $post_id )
```

Replace with:

```
'shop_subscription' === WC_Data_Store::load( 'subscription' )->get_order_type( $id )
```

#### Updating `get_post_type( $product_id )`

I could only find one instance where we're using `get_post_type()` with product ID inside `user_can_not_delete_subscription()`

```
0 === strpos( get_post_type( $args[2] ), 'product' )
```

The issue with`WC_Data_Store::load( 'product' )->get_product_type( $id )` is that it formats product types i.e. `'product_variation'` is returned as just `'variation'` so there's not a simple one-to-one replacement for this one.

We can probably just ignore this one case for now.

**Comment by Jinksi** [comment]
In #153 you mention ignoring the following. Do you think this is still the case?

### Code to ignore
 - [`WC_Subscriptions_Synchroniser::get_sign_up_fee()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/class-wc-subscriptions-synchroniser.php#L1531)
   - Deprecated in 2.x and will be removed
 - [`WCS_Cached_Data_Manager::purge_delete()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/2a0654bf0385cc3e5b097943aea5589e843b05d0/includes/class-wcs-cached-data-manager.php#L73)
   - Deprecated in 2.3

**Comment by Jinksi** [comment]
I attempted a quick and broad change for this issue (replacing instances of `get_post_type` with little attention made to surrounding logic), but it seemed like a risky approach.

Some functions that call `get_post_type` contain other logic that interacts directly with CPT or the DB and they might need significant refactoring and testing or deprecation.

For example `WCS_Download_Handler::repair_permission_data`:

https://github.com/Automattic/woocommerce-subscriptions-core/blob/9e0566089006283af18a70d6326252e5cc0b7219/includes/class-wcs-download-handler.php#L148-L165

**Comment by haszari** [comment]
> I attempted a quick and broad change for this issue (replacing instances of get_post_type with little attention made to surrounding logic), but it seemed like a risky approach.

Good investigation @Jinksi 🙌 

@mattallan  do you think we should do an audit/research spike of these lines of code and then log appropriate fixes as separate issues, detailing the impacted flows? I.e. convert this issue to focus on discovery/research spike.

**Comment by mattallan** [comment]
> Some functions that call get_post_type contain other logic that interacts directly with CPT or the DB and they might need significant refactoring and testing or deprecation.

Thanks @Jinksi for raising this! The purpose of this issue is solely to replace the `get_post_type()` uses with the direct Data Store equivalent. I think any surrounding code can be worked on in a separate issue so feel free to open any issue for things that you find. That said, the function `repair_permission_data()` looks fine to me.

The direct database query is against the `woocommerce_downloadable_product_permissions` table which isn't affected by CPT/HPOS so that will continue to work as per usual and doesn't need to be updated.

> do you think we should do an audit/research spike of these lines of code

I don't think an audit/research spike is necessary here.

`WC_Data_Store::load( 'subscription' )->get_order_type( $id )` is just a [wrapper function for calling `get_post_type()`](https://github.com/woocommerce/woocommerce/blob/4770123fa3a71cf4c561fa6149685c67d0778ee4/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L872-L874) so there isn't any logic changing here. That's mainly why I'm happy for us to change all instances of this function being used in a single PR (similar to the previous PR we made for replacing all `get_post_type()` for 'shop_order').
As for testing this PR, testing one key flow is fixed with these changes, and paying close attention during code review should be sufficient.

There's currently a broken renewal flow caused by `wcs_is_subscription()` still using `get_post_type()` and because these changes are straightforward I figured it would've been best to get them sorted in one PR.

> In https://github.com/Automattic/woocommerce-subscriptions-core/issues/153 you mention ignoring the following. Do you think this is still the case?

I'm happy to leave it up to you :)
I was thinking we just update them all. There's no harm in fixing these up as well (even if we end up removing these older deprecated functions later on)
